e2e/playground.py

```diff
@@ -0,0 +1,2135 @@
+    # Testing the HTML seems like too much.
+    # Testing the text seems like it is not useful.
+
+    # TODO-barret; All methods below:
+    # Could maybe use better formats for the values, but am going to leave as `str` for now
+
+    def expect_min_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-min", value=value, timeout=timeout)
+
+    def expect_max_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-max", value=value, timeout=timeout)
+
+    # def expect_from_to_have_value(
+    #     self, value: AttrValue, *, timeout: Timeout = None
+    # ) -> None:
+    #     expect_attr(self.loc, "data-from", value=value, timeout=timeout)
+
+    def expect_step_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-step", value=value, timeout=timeout)
+
+    def expect_ticks_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-grid", value=value, timeout=timeout)
+
+    def expect_sep_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-prettify-separator", value=value, timeout=timeout)
+
+    def expect_pre_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-prefix", value=value, timeout=timeout)
+
+    def expect_post_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-postfix", value=value, timeout=timeout)
+
+    # def expect_data_type_to_have_value(
+    #     self, value: AttrValue, *, timeout: Timeout = None
+    # ) -> None:
+    #     expect_attr(self.loc, "data-data-type", value=value, timeout=timeout)
+
+    def expect_time_format_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-time-format", value=value, timeout=timeout)
+
+    def expect_timezone_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-timezone", value=value, timeout=timeout)
+
+    def expect_drag_range_to_have_value(
+        self, value: AttrValue, *, timeout: Timeout = None
+    ) -> None:
+        expect_attr(self.loc, "data-drag-interval", value=value, timeout=timeout)
+
+
+def _date_str(date: typing.Union[datetime.date, AttrValue]) -> OptionalStr:
+    if date is None:
+        return None
+    elif isinstance(date, typing.Pattern):
+        return date.pattern
+    elif isinstance(date, datetime.date):
+        return str(date)
+    else:
+        return str(datetime.date.fromisoformat(date))
+
+
+class _DateBase(_WidthContainer, _InputWithLabel):
```

Comment:
`_DateBase` should be a mixin that doesn't have `_WidthContainer` and `_InputWithLabel`, since its use in `DateRangeInput` doesn't have either of those. Introduce a second subclass for use within `DateRangeInput`.